// streamr-client/src/utils/GraphQLClient.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphQLClient = void 0;
const tsyringe_1 = require("tsyringe");
const index_1 = require("./index");
const Config_1 = require("../Config");
const Context_1 = require("./Context");
const HttpFetcher_1 = require("./HttpFetcher");
let GraphQLClient = class GraphQLClient {
    constructor(context, httpFetcher, config) {
        this.httpFetcher = httpFetcher;
        this.config = config;
        this.debug = context.debug.extend((0, index_1.instanceId)(this));
    }
    async sendQuery(gqlQuery) {
        this.debug('GraphQL query: %s', gqlQuery);
        const res = await this.httpFetcher.fetch(this.config.theGraphUrl, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                accept: '*/*',
            },
            body: gqlQuery
        });
        const resText = await res.text();
        let resJson;
        try {
            resJson = JSON.parse(resText);
        }
        catch {
            throw new Error(`GraphQL query failed with "${resText}", check that your theGraphUrl="${this.config.theGraphUrl}" is correct`);
        }
        this.debug('GraphQL response: %o', resJson);
        if (!resJson.data) {
            if (resJson.errors && resJson.errors.length > 0) {
                throw new Error('GraphQL query failed: ' + JSON.stringify(resJson.errors.map((e) => e.message)));
            }
            else {
                throw new Error('GraphQL query failed');
            }
        }
        return resJson.data;
    }
    async *fetchPaginatedResults(createQuery, pageSize = 1000) {
        let lastResultSet;
        do {
            const lastId = (lastResultSet !== undefined) ? lastResultSet[lastResultSet.length - 1].id : '';
            const query = createQuery(lastId, pageSize);
            // eslint-disable-next-line no-await-in-loop
            const response = await this.sendQuery(query);
            const rootKey = Object.keys(response)[0]; // there is a always a one root level property, e.g. "streams" or "permissions"
            const items = response[rootKey];
            yield* items;
            lastResultSet = items;
        } while (lastResultSet.length === pageSize);
    }
    static createWhereClause(variables) {
        const parameterList = Object.keys(variables)
            .filter((k) => variables[k] !== undefined)
            .map((k) => k + ': $' + k)
            .join(' ');
        return `where: { ${parameterList} }`;
    }
};
GraphQLClient = __decorate([
    (0, tsyringe_1.scoped)(tsyringe_1.Lifecycle.ContainerScoped),
    __param(1, (0, tsyringe_1.inject)(HttpFetcher_1.HttpFetcher)),
    __param(2, (0, tsyringe_1.inject)(Config_1.Config.Root)),
    __metadata("design:paramtypes", [Context_1.Context,
        HttpFetcher_1.HttpFetcher, Object])
], GraphQLClient);
exports.GraphQLClient = GraphQLClient;
