// streamr-client/src/Config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULTS = exports.Config = void 0;
/**
 * New Brubeck Configuration.
 * Old Config in ConfigBase.
 * TODO: Disolve ConfigBase.
 */
require("reflect-metadata");
const ConfigBase_1 = __importDefault(require("./ConfigBase"));
const cloneDeep_1 = __importDefault(require("lodash/cloneDeep"));
const merge_1 = __importDefault(require("lodash/merge"));
/**
 * DI Injection tokens for pieces of config.
 * tsyringe needs a concrete value to use as the injection token.
 * In the case of interfaces & types, these have no runtime value
 * so we have to introduce some token to use for their injection.
 * These symbols represent subsections of the full config.
 *
 * For example:
 * config.ethereum can be injected with a token like: @inject(Config.Ethereum)
 */
const BrubeckConfigInjection = {
    Root: Symbol('Config.Root'),
    Auth: Symbol('Config.Auth'),
    Ethereum: Symbol('Config.Ethereum'),
    Network: Symbol('Config.Network'),
    Connection: Symbol('Config.Connection'),
    Subscribe: Symbol('Config.Subscribe'),
    Publish: Symbol('Config.Publish'),
    Cache: Symbol('Config.Cache'),
    StorageNodeRegistry: Symbol('Config.StorageNodeRegistry'),
    Encryption: Symbol('Config.Encryption'),
};
exports.Config = BrubeckConfigInjection;
__exportStar(require("./ConfigBase"), exports);
// TODO: Production values
const BRUBECK_CLIENT_DEFAULTS = {
    debug: {
        inspectOpts: {
            depth: 5,
            maxStringLength: 512
        }
    },
    network: {
        trackers: [
            {
                id: '0xFBB6066c44bc8132bA794C73f58F391273E3bdA1',
                ws: 'wss://brubeck3.streamr.network:30401',
                http: 'https://brubeck3.streamr.network:30401'
            },
            {
                id: '0x3D61bFeFA09CEAC1AFceAA50c7d79BE409E1ec24',
                ws: 'wss://brubeck3.streamr.network:30402',
                http: 'https://brubeck3.streamr.network:30402'
            },
            {
                id: '0xE80FB5322231cBC1e761A0F896Da8E0CA2952A66',
                ws: 'wss://brubeck3.streamr.network:30403',
                http: 'https://brubeck3.streamr.network:30403'
            },
            {
                id: '0xf626285C6AACDE39ae969B9Be90b1D9855F186e0',
                ws: 'wss://brubeck3.streamr.network:30404',
                http: 'https://brubeck3.streamr.network:30404'
            },
            {
                id: '0xce88Da7FE0165C8b8586aA0c7C4B26d880068219',
                ws: 'wss://brubeck3.streamr.network:30405',
                http: 'https://brubeck3.streamr.network:30405'
            },
            {
                id: '0x05e7a0A64f88F84fB1945a225eE48fFC2c48C38E',
                ws: 'wss://brubeck4.streamr.network:30401',
                http: 'https://brubeck4.streamr.network:30401'
            },
            {
                id: '0xF15784106ACd35b0542309CDF2b35cb5BA642C4F',
                ws: 'wss://brubeck4.streamr.network:30402',
                http: 'https://brubeck4.streamr.network:30402'
            },
            {
                id: '0x77FA7Af34108abdf8e92B8f4C4AeC7CbfD1d6B09',
                ws: 'wss://brubeck4.streamr.network:30403',
                http: 'https://brubeck4.streamr.network:30403'
            },
            {
                id: '0x7E83e0bdAF1eF06F31A02f35A07aFB48179E536B',
                ws: 'wss://brubeck4.streamr.network:30404',
                http: 'https://brubeck4.streamr.network:30404'
            },
            {
                id: '0x2EeF37180691c75858Bf1e781D13ae96943Dd388',
                ws: 'wss://brubeck4.streamr.network:30405',
                http: 'https://brubeck4.streamr.network:30405'
            }
        ],
        acceptProxyConnections: false
    },
};
exports.DEFAULTS = BRUBECK_CLIENT_DEFAULTS;
function BrubeckConfig(config) {
    var _a, _b;
    const clonedConfig = (0, cloneDeep_1.default)(config);
    const defaults = (0, cloneDeep_1.default)(BRUBECK_CLIENT_DEFAULTS);
    const userConfig = (0, ConfigBase_1.default)(clonedConfig);
    const result = {
        ...defaults,
        ...userConfig,
        network: {
            ...(0, merge_1.default)(defaults.network || {}, clonedConfig.network),
            trackers: (_b = (_a = clonedConfig.network) === null || _a === void 0 ? void 0 : _a.trackers) !== null && _b !== void 0 ? _b : defaults.network.trackers,
        },
        debug: (0, merge_1.default)(defaults.debug || {}, clonedConfig.debug),
    };
    return result;
}
exports.default = BrubeckConfig;
