// streamr-client/src/DestroySignal.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DestroySignal = void 0;
/**
 * Client-wide destroy signal.
 */
const tsyringe_1 = require("tsyringe");
const utils_1 = require("./utils");
const Context_1 = require("./utils/Context");
const Signal_1 = __importDefault(require("./utils/Signal"));
/**
 * Listen to onDestroy to fire cleanup code on destroy.
 * Careful not to introduce memleaks.
 * Trigger this to destroy the client.
 */
let DestroySignal = class DestroySignal {
    constructor(context) {
        this.onDestroy = Signal_1.default.once();
        this.trigger = this.destroy;
        this.id = (0, utils_1.instanceId)(this);
        this.debug = context.debug.extend(this.id);
        this.onDestroy(() => {
            this.debug('triggered');
        });
    }
    destroy() {
        return this.onDestroy.trigger();
    }
    assertNotDestroyed(context, msg = 'Client is destroyed. Create a new instance') {
        if (this.isDestroyed()) {
            throw new Context_1.ContextError(context, msg);
        }
    }
    isDestroyed() {
        return this.onDestroy.triggerCount() > 0;
    }
};
DestroySignal = __decorate([
    (0, tsyringe_1.scoped)(tsyringe_1.Lifecycle.ContainerScoped),
    __metadata("design:paramtypes", [Context_1.Context])
], DestroySignal);
exports.DestroySignal = DestroySignal;
