// streamr-client/src/index-exports.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Contract = exports.BigNumber = exports.validateConfig = exports.ConfigTest = exports.getTrackerRegistryFromContract = exports.StreamPermission = exports.ResendSubscription = exports.Subscription = void 0;
/**
 * This file captures named exports so we can manipulate them for cjs/browser builds.
 */
__exportStar(require("./StreamrClient"), exports);
__exportStar(require("./Config"), exports);
__exportStar(require("./Stream"), exports);
__exportStar(require("./encryption/Encryption"), exports);
var Subscription_1 = require("./subscribe/Subscription");
Object.defineProperty(exports, "Subscription", { enumerable: true, get: function () { return Subscription_1.Subscription; } });
var ResendSubscription_1 = require("./subscribe/ResendSubscription");
Object.defineProperty(exports, "ResendSubscription", { enumerable: true, get: function () { return ResendSubscription_1.ResendSubscription; } });
var permission_1 = require("./permission");
Object.defineProperty(exports, "StreamPermission", { enumerable: true, get: function () { return permission_1.StreamPermission; } });
__exportStar(require("./LoginEndpoints"), exports);
__exportStar(require("./StreamEndpoints"), exports);
var getTrackerRegistryFromContract_1 = require("./getTrackerRegistryFromContract");
Object.defineProperty(exports, "getTrackerRegistryFromContract", { enumerable: true, get: function () { return getTrackerRegistryFromContract_1.getTrackerRegistryFromContract; } });
const ConfigTest_1 = __importDefault(require("./ConfigTest"));
exports.ConfigTest = ConfigTest_1.default;
const ConfigBase_1 = require("./ConfigBase");
Object.defineProperty(exports, "validateConfig", { enumerable: true, get: function () { return ConfigBase_1.validateConfig; } });
__exportStar(require("./dataunion/DataUnion"), exports);
__exportStar(require("./authFetch"), exports);
__exportStar(require("./types"), exports);
// TODO should export these to support StreamMessageAsObject:
// export {
//   StreamMessageType, ContentType, EncryptionType, SignatureType
// } from 'streamr-client-protocol/dist/src/protocol/message_layer/StreamMessage'
var bignumber_1 = require("@ethersproject/bignumber");
Object.defineProperty(exports, "BigNumber", { enumerable: true, get: function () { return bignumber_1.BigNumber; } });
var contracts_1 = require("@ethersproject/contracts");
Object.defineProperty(exports, "Contract", { enumerable: true, get: function () { return contracts_1.Contract; } });
