// streamr-client/src/StreamIDBuilder.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamIDBuilder = exports.formStreamDefinitionDescription = exports.DEFAULT_PARTITION = void 0;
const streamr_client_protocol_1 = require("streamr-client-protocol");
const Ethereum_1 = __importDefault(require("./Ethereum"));
const tsyringe_1 = require("tsyringe");
exports.DEFAULT_PARTITION = 0;
/* eslint-disable no-else-return */
function pickStreamId(definition) {
    const obj = definition;
    if (obj.id !== undefined) {
        return obj.id;
    }
    else if (obj.stream !== undefined) {
        return obj.stream;
    }
    else if (obj.streamId !== undefined) {
        return obj.streamId;
    }
    else {
        throw new Error('streamDefinition: object must have property: "id", "stream", or "streamId"');
    }
}
function parseRawDefinition(definition) {
    if (typeof definition === 'string') {
        return streamr_client_protocol_1.StreamPartIDUtils.parseRawElements(definition);
    }
    else if (typeof definition === 'object') {
        return [pickStreamId(definition), definition.partition];
    }
    else {
        throw new Error('streamDefinition: must be of type string or object');
    }
}
/*
 * Use this e.g. for internal logging purposes where any kind of string representation is fine.
 *
 * Note that sometimes it is better to write this object to a log than just write e.g. JSON
 * representation of a stream definition. The stream definition object may be any instance which
 * implements the StreamDefinition interface (e.g a Stream object), and we don't want to write
 * many extra fields to the log. By converting the streamDefinition to a string with this method
 * we write just the information we need. Also calling JSON.strinfigy for a Stream object may
 * fail because the object contains circular refences.
 * */
const formStreamDefinitionDescription = (definition) => {
    const raw = parseRawDefinition(definition);
    let str = raw[0];
    if (raw[1] !== undefined) {
        str += '#' + raw[1]; // TODO '#' from a constant?
    }
    return str;
};
exports.formStreamDefinitionDescription = formStreamDefinitionDescription;
/* eslint-disable no-else-return */
let StreamIDBuilder = class StreamIDBuilder {
    constructor(ethereum) {
        this.ethereum = ethereum;
    }
    async toStreamID(streamIdOrPath) {
        let address;
        if (streamr_client_protocol_1.StreamIDUtils.isPathOnlyFormat(streamIdOrPath) && this.ethereum.isAuthenticated()) {
            address = await this.ethereum.getAddress();
        }
        return (0, streamr_client_protocol_1.toStreamID)(streamIdOrPath, address);
    }
    async toStreamPartID(definition) {
        const [streamId, streamPartition] = await this.toStreamPartElements(definition);
        return (0, streamr_client_protocol_1.toStreamPartID)(streamId, streamPartition !== null && streamPartition !== void 0 ? streamPartition : exports.DEFAULT_PARTITION);
    }
    async toStreamPartElements(definition) {
        const [streamId, streamPartition] = parseRawDefinition(definition);
        return [await this.toStreamID(streamId), streamPartition];
    }
    async match(definition, streamPartId) {
        const [targetStreamId, targetPartition] = await this.toStreamPartElements(definition);
        return targetStreamId === streamr_client_protocol_1.StreamPartIDUtils.getStreamID(streamPartId)
            && (targetPartition === undefined || targetPartition === streamr_client_protocol_1.StreamPartIDUtils.getStreamPartition(streamPartId));
    }
};
StreamIDBuilder = __decorate([
    (0, tsyringe_1.scoped)(tsyringe_1.Lifecycle.ContainerScoped),
    __param(0, (0, tsyringe_1.inject)(Ethereum_1.default)),
    __metadata("design:paramtypes", [Ethereum_1.default])
], StreamIDBuilder);
exports.StreamIDBuilder = StreamIDBuilder;
