// streamr-network/dist/src/connection/PeerInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PeerInfo = exports.PeerType = void 0;
const streamr_client_protocol_1 = require("streamr-client-protocol");
var PeerType;
(function (PeerType) {
    PeerType["Tracker"] = "tracker";
    PeerType["Node"] = "node";
    PeerType["Unknown"] = "unknown";
})(PeerType = exports.PeerType || (exports.PeerType = {}));
const defaultControlLayerVersions = streamr_client_protocol_1.ControlLayer.ControlMessage.getSupportedVersions();
const defaultMessageLayerVersions = streamr_client_protocol_1.MessageLayer.StreamMessage.getSupportedVersions();
class PeerInfo {
    constructor(peerId, peerType, controlLayerVersions, messageLayerVersions, peerName, location) {
        if (!peerId) {
            throw new Error('peerId not given');
        }
        if (!peerType) {
            throw new Error('peerType not given');
        }
        if (!Object.values(PeerType).includes(peerType)) {
            throw new Error(`peerType ${peerType} not in peerTypes list`);
        }
        if (!controlLayerVersions || controlLayerVersions.length === 0) {
            throw new Error('controlLayerVersions not given');
        }
        if (!messageLayerVersions || messageLayerVersions.length === 0) {
            throw new Error('messageLayerVersions not given');
        }
        this.peerId = peerId;
        this.peerType = peerType;
        this.controlLayerVersions = controlLayerVersions;
        this.messageLayerVersions = messageLayerVersions;
        this.peerName = peerName ? peerName : null;
        this.location = location;
    }
    static newTracker(peerId, peerName, controlLayerVersions, messageLayerVersions, location) {
        return new PeerInfo(peerId, PeerType.Tracker, controlLayerVersions || defaultControlLayerVersions, messageLayerVersions || defaultMessageLayerVersions, peerName, location);
    }
    static newNode(peerId, peerName, controlLayerVersions, messageLayerVersions, location) {
        return new PeerInfo(peerId, PeerType.Node, controlLayerVersions || defaultControlLayerVersions, messageLayerVersions || defaultMessageLayerVersions, peerName, location);
    }
    static newUnknown(peerId) {
        return new PeerInfo(peerId, PeerType.Unknown, defaultControlLayerVersions, defaultMessageLayerVersions);
    }
    static fromObject({ peerId, peerType, peerName, location, controlLayerVersions, messageLayerVersions }) {
        return new PeerInfo(peerId, peerType, controlLayerVersions || defaultControlLayerVersions, messageLayerVersions || defaultMessageLayerVersions, peerName, location !== null && location !== void 0 ? location : undefined);
    }
    isTracker() {
        return this.peerType === PeerType.Tracker;
    }
    isNode() {
        return this.peerType === PeerType.Node;
    }
    toString() {
        return (this.peerName ? `${this.peerName}` : '') + `<${this.peerId.slice(0, 8)}>`;
    }
}
exports.PeerInfo = PeerInfo;
