// streamr-client/src/subscribe/SubscriptionSession.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const tsyringe_1 = require("tsyringe");
const utils_1 = require("../utils");
const Context_1 = require("../utils/Context");
const Signal_1 = __importDefault(require("../utils/Signal"));
const MessageStream_1 = require("./MessageStream");
const SubscribePipeline_1 = __importDefault(require("./SubscribePipeline"));
const Container_1 = require("../Container");
const BrubeckNode_1 = __importDefault(require("../BrubeckNode"));
/**
 * Manages adding & removing subscriptions to node as needed.
 * A session contains one or more subscriptions to a single streamId + streamPartition pair.
 */
let SubscriptionSession = class SubscriptionSession {
    constructor(context, streamPartId, container) {
        /** active subs */
        this.subscriptions = new Set();
        this.pendingRemoval = new WeakSet();
        this.isRetired = false;
        this.isStopped = false;
        this.onRetired = Signal_1.default.once();
        this.onMessageInput = async (msg) => {
            if (!msg || this.isStopped || this.isRetired) {
                return;
            }
            if (msg.getStreamPartID() !== this.streamPartId) {
                return;
            }
            await this.pipeline.push(msg);
        };
        this.updateNodeSubscriptions = (() => {
            let node;
            return (0, utils_1.Scaffold)([
                async () => {
                    node = await this.subscribe();
                    return async () => {
                        const prevNode = node;
                        node = undefined;
                        await this.unsubscribe(prevNode);
                        await this.stop();
                    };
                },
            ], () => this.shouldBeSubscribed());
        })();
        this.id = (0, utils_1.instanceId)(this);
        this.debug = context.debug.extend(this.id);
        this.streamPartId = streamPartId;
        this.distributeMessage = this.distributeMessage.bind(this);
        this.node = container.resolve(BrubeckNode_1.default);
        this.onError = this.onError.bind(this);
        this.pipeline = (0, SubscribePipeline_1.default)(new MessageStream_1.MessageStream(this), this.streamPartId, this, container)
            .onError(this.onError)
            .pipe(this.distributeMessage)
            .onBeforeFinally(async () => {
            if (!this.isStopped) {
                await this.stop();
            }
        });
        this.pipeline.flow();
    }
    async retire() {
        if (this.isRetired) {
            return;
        }
        this.isRetired = true;
        await this.onRetired.trigger();
    }
    async onError(error) {
        await Promise.allSettled([...this.subscriptions].map(async (sub) => {
            await sub.handleError(error);
        }));
    }
    async *distributeMessage(src) {
        for await (const msg of src) {
            await Promise.all([...this.subscriptions].map(async (sub) => {
                await sub.push(msg);
            }));
            yield msg;
        }
    }
    async subscribe() {
        this.debug('subscribe');
        const node = await this.node.getNode();
        node.addMessageListener(this.onMessageInput);
        node.subscribe(this.streamPartId);
        return node;
    }
    async unsubscribe(node) {
        this.debug('unsubscribe');
        this.pipeline.end();
        this.pipeline.return();
        this.pipeline.onError.end(new Error('done'));
        node.removeMessageListener(this.onMessageInput);
        node.unsubscribe(this.streamPartId);
    }
    async updateSubscriptions() {
        await this.updateNodeSubscriptions();
        if (!this.shouldBeSubscribed() && !this.isStopped) {
            await this.stop();
        }
    }
    shouldBeSubscribed() {
        return !this.isRetired && !this.isStopped && !!this.count();
    }
    async stop() {
        this.debug('stop');
        this.isStopped = true;
        this.pipeline.end();
        await this.retire();
        await this.pipeline.return();
    }
    has(sub) {
        return this.subscriptions.has(sub);
    }
    async waitForNeighbours(numNeighbours = 1, timeout = 10000) {
        return (0, utils_1.until)(async () => {
            if (!this.shouldBeSubscribed()) {
                return true;
            } // abort
            const node = await this.node.getNode();
            if (!this.shouldBeSubscribed()) {
                return true;
            } // abort
            return node.getNeighborsForStreamPart(this.streamPartId).length >= numNeighbours;
        }, timeout);
    }
    /**
     * Add subscription & appropriate connection handle.
     */
    async add(sub) {
        if (!sub || this.subscriptions.has(sub) || this.pendingRemoval.has(sub)) {
            return;
        } // already has
        this.debug('add', sub.id);
        this.subscriptions.add(sub);
        sub.onBeforeFinally(() => {
            return this.remove(sub);
        });
        await this.updateSubscriptions();
    }
    /**
     * Remove subscription & appropriate connection handle.
     */
    async remove(sub) {
        if (!sub || this.pendingRemoval.has(sub) || !this.subscriptions.has(sub)) {
            return;
        }
        this.debug('remove >>', sub.id);
        this.pendingRemoval.add(sub);
        this.subscriptions.delete(sub);
        try {
            if (!sub.isUnsubscribed && !sub.isDone()) {
                await sub.unsubscribe();
            }
        }
        finally {
            await this.updateSubscriptions();
            this.debug('remove <<', sub.id);
        }
    }
    /**
     * Remove all subscriptions & subscription connection handles
     */
    async removeAll() {
        this.debug('removeAll %d', this.count());
        await Promise.all([...this.subscriptions].map((sub) => (this.remove(sub))));
    }
    /**
     * How many subscriptions
     */
    count() {
        return this.subscriptions.size;
    }
};
SubscriptionSession = __decorate([
    __param(2, (0, tsyringe_1.inject)(Container_1.BrubeckContainer)),
    __metadata("design:paramtypes", [Context_1.Context, String, Object])
], SubscriptionSession);
exports.default = SubscriptionSession;
