// streamr-network/dist/src/protocol/NodeToNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeToNode = exports.Event = void 0;
const events_1 = require("events");
const streamr_client_protocol_1 = require("streamr-client-protocol");
const Logger_1 = require("../helpers/Logger");
const utils_1 = require("./utils");
const IWebRtcEndpoint_1 = require("../connection/IWebRtcEndpoint");
var Event;
(function (Event) {
    Event["NODE_CONNECTED"] = "streamr:node-node:node-connected";
    Event["NODE_DISCONNECTED"] = "streamr:node-node:node-disconnected";
    Event["DATA_RECEIVED"] = "streamr:node-node:stream-data";
    Event["LOW_BACK_PRESSURE"] = "streamr:node-node:low-back-pressure";
    Event["HIGH_BACK_PRESSURE"] = "streamr:node-node:high-back-pressure";
    Event["PUBLISH_STREAM_REQUEST_RECEIVED"] = "node-node:publish-only-stream-request-received";
    Event["PUBLISH_STREAM_RESPONSE_RECEIVED"] = "node-node:publish-only-stream-response-received";
    Event["LEAVE_REQUEST_RECEIVED"] = "node-node:leave-request-received";
})(Event = exports.Event || (exports.Event = {}));
const eventPerType = {};
eventPerType[streamr_client_protocol_1.ControlLayer.ControlMessage.TYPES.BroadcastMessage] = Event.DATA_RECEIVED;
eventPerType[streamr_client_protocol_1.ControlLayer.ControlMessage.TYPES.PublishStreamConnectionRequest] = Event.PUBLISH_STREAM_REQUEST_RECEIVED;
eventPerType[streamr_client_protocol_1.ControlLayer.ControlMessage.TYPES.PublishStreamConnectionResponse] = Event.PUBLISH_STREAM_RESPONSE_RECEIVED;
eventPerType[streamr_client_protocol_1.ControlLayer.ControlMessage.TYPES.UnsubscribeRequest] = Event.LEAVE_REQUEST_RECEIVED;
class NodeToNode extends events_1.EventEmitter {
    constructor(endpoint) {
        super();
        this.endpoint = endpoint;
        endpoint.on(IWebRtcEndpoint_1.Event.PEER_CONNECTED, (peerInfo) => this.onPeerConnected(peerInfo));
        endpoint.on(IWebRtcEndpoint_1.Event.PEER_DISCONNECTED, (peerInfo) => this.onPeerDisconnected(peerInfo));
        endpoint.on(IWebRtcEndpoint_1.Event.MESSAGE_RECEIVED, (peerInfo, message) => this.onMessageReceived(peerInfo, message));
        endpoint.on(IWebRtcEndpoint_1.Event.LOW_BACK_PRESSURE, (peerInfo) => this.onLowBackPressure(peerInfo));
        endpoint.on(IWebRtcEndpoint_1.Event.HIGH_BACK_PRESSURE, (peerInfo) => this.onHighBackPressure(peerInfo));
        this.logger = new Logger_1.Logger(module);
    }
    connectToNode(receiverNodeId, trackerAddress, trackerInstructed = true) {
        return this.endpoint.connect(receiverNodeId, trackerAddress, trackerInstructed);
    }
    sendData(receiverNodeId, streamMessage) {
        return this.send(receiverNodeId, new streamr_client_protocol_1.ControlLayer.BroadcastMessage({
            requestId: '',
            streamMessage,
        }));
    }
    send(receiverNodeId, message) {
        const [controlLayerVersion, messageLayerVersion] = this.getNegotiatedProtocolVersionsOnNode(receiverNodeId);
        return this.endpoint.send(receiverNodeId, message.serialize(controlLayerVersion, messageLayerVersion)).then(() => message);
    }
    disconnectFromNode(receiverNodeId, reason) {
        this.endpoint.close(receiverNodeId, reason);
    }
    /**
     * @deprecated
     */
    getAddress() {
        return this.endpoint.getAddress();
    }
    stop() {
        this.endpoint.stop();
    }
    onPeerConnected(peerInfo) {
        if (peerInfo.isNode()) {
            this.emit(Event.NODE_CONNECTED, peerInfo.peerId);
        }
    }
    onPeerDisconnected(peerInfo) {
        if (peerInfo.isNode()) {
            this.emit(Event.NODE_DISCONNECTED, peerInfo.peerId);
        }
    }
    onMessageReceived(peerInfo, rawMessage) {
        if (peerInfo.isNode()) {
            const message = (0, utils_1.decode)(rawMessage, streamr_client_protocol_1.ControlLayer.ControlMessage.deserialize);
            if (message != null) {
                this.emit(eventPerType[message.type], message, peerInfo.peerId);
            }
            else {
                this.logger.warn('invalid message from %s: %s', peerInfo, rawMessage);
            }
        }
    }
    onLowBackPressure(peerInfo) {
        if (peerInfo.isNode()) {
            this.emit(Event.LOW_BACK_PRESSURE, peerInfo.peerId);
        }
    }
    onHighBackPressure(peerInfo) {
        if (peerInfo.isNode()) {
            this.emit(Event.HIGH_BACK_PRESSURE, peerInfo.peerId);
        }
    }
    getRtts() {
        return this.endpoint.getRtts();
    }
    getNegotiatedProtocolVersionsOnNode(nodeId) {
        const messageLayerVersion = this.endpoint.getNegotiatedMessageLayerProtocolVersionOnNode(nodeId)
            || this.endpoint.getDefaultMessageLayerProtocolVersion();
        const controlLayerVersion = this.endpoint.getNegotiatedControlLayerProtocolVersionOnNode(nodeId)
            || this.endpoint.getDefaultControlLayerProtocolVersion();
        return [controlLayerVersion, messageLayerVersion];
    }
    async requestPublishOnlyStreamConnection(nodeId, streamPartId) {
        const [streamId, streamPartition] = streamr_client_protocol_1.StreamPartIDUtils.getStreamIDAndPartition(streamPartId);
        await this.send(nodeId, new streamr_client_protocol_1.ControlLayer.PublishStreamConnectionRequest({
            requestId: '',
            senderId: nodeId,
            streamId,
            streamPartition
        }));
    }
    async leaveStreamOnNode(nodeId, streamPartId) {
        const [streamId, streamPartition] = streamr_client_protocol_1.StreamPartIDUtils.getStreamIDAndPartition(streamPartId);
        await this.send(nodeId, new streamr_client_protocol_1.ControlLayer.UnsubscribeRequest({
            requestId: '',
            streamId,
            streamPartition
        }));
    }
    async respondToPublishOnlyStreamConnectionRequest(nodeId, streamPartId, accepted) {
        const [streamId, streamPartition] = streamr_client_protocol_1.StreamPartIDUtils.getStreamIDAndPartition(streamPartId);
        await this.send(nodeId, new streamr_client_protocol_1.ControlLayer.PublishStreamConnectionResponse({
            requestId: '',
            senderId: nodeId,
            streamId,
            streamPartition,
            accepted
        }));
    }
    getAllConnectionNodeIds() {
        return this.endpoint.getAllConnectionNodeIds();
    }
}
exports.NodeToNode = NodeToNode;
